pandas/core/arrays/datetimelike.py

```diff
@@ -366,15 +460,109 @@ def __getitem__(self, key):
 
         return self._simple_new(result, **attribs)
 
+    def __setitem__(
+            self,
+            key,    # type: Union[int, Sequence[int], Sequence[bool], slice]
+            value,  # type: Union[NaTType, Scalar, Sequence[Scalar]]
+    ):
+        # type: (...) -> None
+        # I'm fudging the types a bit here. The "Scalar" above really depends
+        # on type(self). For PeriodArray, it's Period (or stuff coercible
+        # to a period in from_sequence). For DatetimeArray, it's Timestamp...
+        # I don't know if mypy can do that, possibly with Generics.
+        # https://mypy.readthedocs.io/en/latest/generics.html
+
+        # n.b. This is moved from PeriodArray with the following changes
+        # 1. changed dedicated ctor (period_array) to _from_sequence
+        # 2. Changed freq checking to use `_check_compatible_with`
+        # 3. Handle `value=iNaT` (may be able to revert. Check internals.)
+        if is_list_like(value):
+            is_slice = isinstance(key, slice)
+            if (not is_slice
+                    and len(key) != len(value)
+                    and not com.is_bool_indexer(key)):
+                msg = ("shape mismatch: value array of length '{}' does not "
+                       "match indexing result of length '{}'.")
+                raise ValueError(msg.format(len(key), len(value)))
+            if not is_slice and len(key) == 0:
+                return
+
+            value = type(self)._from_sequence(value, dtype=self.dtype)
+            self._check_compatible_with(value)
+            value = value.asi8
+        elif isinstance(value, self._scalar_type):
+            self._check_compatible_with(value)
+            value = self._unbox_scalar(value)
+        elif isna(value) or value == iNaT:
+            # TODO: Right now DatetimeTZBlock.fill_value is iNaT.
+            # There's some confuction about whether Block.fill_value should
+            # be the NA value or the storage value.
+            value = iNaT
+        else:
+            msg = (
+                "'value' should be a '{scalar}', 'NaT', or array of those. "
+                "Got '{typ}' instead."
+            )
+            raise TypeError(msg.format(scalar=self._scalar_type.__name__,
+                                       typ=type(value).__name__))
+        self._data[key] = value
+
+    def view(self, dtype=None):
```

Comment:
Need to have an API discussion on `DatetimeArrayLike.view` (and ExtensionArray.view more generally). Series.view uses it directly. 

I don't know whether or not `view` would have a meaning for ndarrays that aren't backed by numpy.

Reply:
Having DTA.view has proved to be useful for writing generic code that can handle ndarray / DatetimeIndex / DatetimeArray. I propose that we keep it.